File: snippets/dijkstra-algorithm/dijkstra.cpp
#include <bits/stdc++.h>

using namespace std;

class Graph {
public:
    struct Edge {
        int dest;
        int weight;
        Edge(int d, int w) : dest{d}, weight{w} {}
    };

    int V;
    vector<list<Edge>> adj;

    Graph(int V) {
        this->V = V;
        adj.resize(V);
    }

    void AddEdge(int src, int dest, int weight) {
        Edge e(dest, weight);
        adj[src].push_back(e);
    }

    void AddBiEdge(int u, int v, int weight) {
        AddEdge(u, v, weight);
        AddEdge(v, u, weight);
    }

    pair<vector<int>, vector<int>> Dijkstra(int src) {
        vector<int> dist(V, INT_MAX);
        vector<int> pred(V, -1);
        vector<bool> visited(V, false);

        dist[src] = 0;

        for (int count = 0; count < V - 1; count++) {
            int u = MinDistance(dist, visited);
            visited[u] = true;

            for (const auto& edge : adj[u]) {
                if (!visited[edge.dest]) {
                    int d = dist[u] + edge.weight;
                    if (d < dist[edge.dest]) {
                        dist[edge.dest] = d;
                        pred[edge.dest] = u;
                    }
                }
            }
        }

        return make_pair(dist, pred);  ///
    }

    int MinDistance(const vector<int>& dist, const vector<bool>& visited) {
        int min_dist = INT_MAX, min_vertex = 0;

        for (int v = 0; v < V; v++) {
            if (!visited[v] && dist[v] < min_dist) {
                min_dist = dist[v];
                min_vertex = v;
            }
        }

        return min_vertex;
    }
};

File: snippets/dijkstra-algorithm/dijkstra.py
from collections import *

class Graph(object):
    def __init__(self, V):
        self.Edge = namedtuple('Edge', 'dest weight')
        self.V = V
        self.adj = [[] for i in range(V)]

    def add_edge(self, src, dest, weight):
        e = self.Edge(dest, weight)
        self.adj[src].append(e)

    def add_bi_edge(self, u, v, weight):
        self.add_edge(u, v, weight)
        self.add_edge(v, u, weight)

    def dijkstra(self, src):
        dist = [float('inf') for i in range(self.V)]
        pred = [-1 for i in range(self.V)]
        visited = [False for i in range(self.V)]

        dist[src] = 0

        for count in range(self.V):
            u = self.min_distance(dist, visited)
            visited[u] = True

            for edge in self.adj[u]:
                if not visited[edge.dest]:
                    d = dist[u] + edge.weight
                    if (d < dist[edge.dest]):
                        dist[edge.dest] = d
                        pred[edge.dest] = u

        return (dist, pred)  ###

    def min_distance(self, dist, visited):
        min_dist = float('inf')
        min_vertex = 0

        for v in range(self.V):
            if (dist[v] < min_dist) and (not visited[v]):
                min_dist = dist[v]
                min_vertex = v

        return min_vertex
